Lesson3 (Container adapters)/tree.cpp

```diff
@@ -0,0 +1,315 @@
+#include <iostream>
+#include <queue>
+#include <stdexcept>
+#include "tree.h"
+
+
+
+nodeIterator nodeIterator::end()
+{
+    nodeIterator it(nullptr);
+    return it;
+}
+
+bool nodeIterator::operator!=(const nodeIterator& other) const
+{
+    return (this->container->key != other.container->key);
+}
+
+bool nodeIterator::operator==(const nodeIterator& other) const
+{
+    return (this->container->key == other.container->key);
+}
+nodeIterator nodeIterator::begin()
+{
+    Node* tree = this->container;
+    while(tree->parent != nullptr)
+    {
+        tree = tree->parent;
+    }
+    nodeIterator it(tree);
+    return it;
+
+}
+
+Node* nodeIterator::descent(int count, int count2, int inLoop, Node*& tree, Node*& copyTree)
```

Comment:
why your tree doesn't contain the "count" member? if you will do it with a wrapper, handling will be much easier